From 2ff0096feb132eea95220e6256379603c004b1a8 Mon Sep 17 00:00:00 2001
From: lyintel <yang.a.lu@intel.com>
Date: Tue, 31 May 2022 20:32:20 +0000
Subject: [PATCH] INTERNAL: fix a deadlock during execute CTS cases

This patch try to fix a deadlock issue when runing
android.media.cts.HeifWriterTest#testInputSurface_Grid_Handler
In this case, there are two deadlock threads involved, the HeifWriter
thread and a drawframe thread. The drawframe thread will try to draw a
frame and send it to HeifWriter. By calling st_api_make_currentst() it
will pending in surface dequeubuffer() with hold of EGL lock, and will
block HeifWriter thread executing, the buffer will be returned after
HeifWriter handled, In this case HeifWriter waiting for the EGL lock,
and drawframe thread wait for buffer with EGL lock hold, so deadlock
produced.

In case this is not the first time run of make_current(), allocate
buffer can be postponed, and allocate buffer in a context without
hold of EGL lock.

Tracked-On: OAM-96997
Signed-off-by: Yang, Dong <dong.yang@intel.com>

INTERNEL: allocate buffer if there isn't _ColorReadBuffer

    allocate buffer if there isn't _ColorReadBuffer

Tracked-On: OAM-99287
Signed-off-by: Yang, Dong <dong.yang@intel.com>
---
 src/mesa/state_tracker/st_manager.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/src/mesa/state_tracker/st_manager.c b/src/mesa/state_tracker/st_manager.c
index 6d7c90e17c1..437cd840f2a 100644
--- a/src/mesa/state_tracker/st_manager.c
+++ b/src/mesa/state_tracker/st_manager.c
@@ -1140,9 +1140,11 @@ st_api_make_current(struct st_context *st,
          return false;
 
       if (stdraw && stread) {
-         st_framebuffer_validate(stdraw, st);
-         if (stread != stdraw)
-            st_framebuffer_validate(stread, st);
+         if (st->ctx->FirstTimeCurrent || !stread->_ColorReadBuffer || !stdraw->_ColorReadBuffer) {
+            st_framebuffer_validate(stdraw, st);
+            if (stread != stdraw)
+               st_framebuffer_validate(stread, st);
+         }
 
          ret = _mesa_make_current(st->ctx, stdraw, stread);
 
-- 
2.40.1

